spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/utils/PlanHelper.scala

```diff
@@ -0,0 +1,168 @@
+    }
+  }
+
+  private def getFileFilterPlan(cond: Expression, scanRelation: DataSourceV2ScanRelation): LogicalPlan = {
+    Filter(cond, scanRelation)
+  }
+
+  private def pushDownCondition(cond: Expression,
+                                scanBuilder: ScanBuilder,
+                                output: Seq[AttributeReference]): Unit = {
+    val predicates = splitConjunctivePredicates(cond)
+    val normalizedPredicates = DataSourceStrategy.normalizeExprs(predicates, output)
+    PushDownUtils.pushFilters(scanBuilder, normalizedPredicates)
+  }
+
+  private def buildAggregatePlan(prunedTargetPlan: LogicalPlan): LogicalPlan = {
+    val fileAttr = findOutputAttr(prunedTargetPlan, FILE_NAME_COL)
+    Aggregate(Seq(fileAttr), Seq(fileAttr), prunedTargetPlan)
+  }
+
+  private def buildPlanWithFileAccumulator( fileAccum: SetAccumulator[String],
+                                            prunedTargetPlan: LogicalPlan): LogicalPlan = {
+    val fileAttr = findOutputAttr(prunedTargetPlan, FILE_NAME_COL)
+    val rowIdAttr = findOutputAttr(prunedTargetPlan, ROW_ID_COL)
+    val projectList = Seq(fileAttr, rowIdAttr,
+      Alias(AccumulateFiles(fileAccum, fileAttr), AFFECTED_FILES_ACC_ALIAS_NAME )())
+    val projectPlan = Project(projectList, prunedTargetPlan)
+    val affectedFilesAttr = findOutputAttr(projectPlan, AFFECTED_FILES_ACC_ALIAS_NAME)
+    val aggSumCol =
+      Alias(AggregateExpression(Sum(affectedFilesAttr), Complete, false), SUM_ROW_ID_ALIAS_NAME)()
+    val aggPlan = Aggregate(Seq(rowIdAttr), Seq(aggSumCol), projectPlan)
+    val sumAttr = findOutputAttr(aggPlan, SUM_ROW_ID_ALIAS_NAME)
+    val havingExpr = GreaterThan(sumAttr, Literal(1L))
+    Filter(havingExpr, aggPlan)
+  }
+
+  def findOutputAttr(plan: LogicalPlan, attrName: String): Attribute = {
+    val resolver = SQLConf.get.resolver
+    plan.output.find(attr => resolver(attr.name, attrName)).getOrElse {
+      throw new AnalysisException(s"Cannot find $attrName in ${plan.output}")
+    }
+  }
+
+  def newWriteInfo(schema: StructType): LogicalWriteInfo = {
+    val uuid = UUID.randomUUID()
+    LogicalWriteInfoImpl(queryId = uuid.toString, schema, CaseInsensitiveStringMap.empty)
+  }
+
+  private def toOutputAttrs(schema: StructType, output: Seq[AttributeReference]): Seq[AttributeReference] = {
+    val nameToAttr = output.map(_.name).zip(output).toMap
+    schema.toAttributes.map {
+      a => nameToAttr.get(a.name) match {
+        case Some(ref) =>
+          // keep the attribute id if it was present in the relation
+          a.withExprId(ref.exprId)
+        case _ =>
+          // if the field is new, create a new attribute
+          AttributeReference(a.name, a.dataType, a.nullable, a.metadata)()
+      }
+    }
+  }
+
+  def toCatalyst(partitionSpec: PartitionSpec, plan: LogicalPlan): Seq[Expression] = {
+    val partitionFields = partitionSpec.fields().asScala.toSeq
+    partitionFields.map { field =>
+      findOutputAttr(plan, field.name())
+    }
+  }
```

Comment:
The conversion for partition spec and sort order needs to handle transforms. And the partition field name is not a column in the Spark schema.